thread.c

```diff
@@ -4434,6 +4434,25 @@ rb_thread_wait_for_single_fd(int fd, int events, struct timeval *timeout)
     int r;
     VALUE ptr = (VALUE)&args;
 
+#ifdef RUBY_THREAD_PTHREAD_H
+    rb_thread_t *th = GET_THREAD();
+    if (!th->nt->dedicated) {
+        rb_hrtime_t rel, *prel;
+
+        if (timeout) {
+            rel = rb_timeval2hrtime(timeout);
+            prel = &rel;
+        }
+        else {
+            prel = NULL;
+        }
+
+        if (thread_sched_wait_events(TH_SCHED(th), th, fd, waitfd_to_waiting_flag(events), prel)) {
+            return 0; // timeout
+        }
+    }
+#endif
```

Comment:
This code is copied right from the other `rb_thread_wait_for_single_fd` method, so maybe it should be in a shared function.